src/rust-crypto/dehydration.ts

```diff
@@ -0,0 +1,262 @@
+    public async isKeyStored(): Promise<boolean> {
+        return Boolean(await this.secretStorage.isStored(SECRET_STORAGE_NAME.name));
+    }
+
+    /** Reset the dehydration key
+     */
+    public async resetKey(): Promise<void> {
+        const key = new Uint8Array(32);
+        crypto.getRandomValues(key);
+        await this.secretStorage.store(SECRET_STORAGE_NAME.name, encodeUnpaddedBase64(key));
+        this.key = key;
+    }
+
+    /** Get and cache the encryption key from secret storage
+     *
+     * If `create` is `true`, creates a new key if no existing key is present.
+     */
+    private async getKey(create: boolean): Promise<boolean> {
+        if (this.key === undefined) {
+            const keyB64 = await this.secretStorage.get(SECRET_STORAGE_NAME.name);
+            if (keyB64 === undefined) {
+                if (!create) {
+                    return false;
+                }
+                await this.resetKey();
+            } else {
+                this.key = decodeBase64(keyB64);
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Rehydrate the dehydrated device stored on the server
+     *
+     * Checks if there is a dehydrated device on the server.  If so, rehydrates
+     * the device and processes the to-device events.
+     *
+     * Returns whether or not a dehydrated device was found.
+     */
+    public async rehydrateDeviceIfAvailable(): Promise<boolean> {
+        if (!(await this.getKey(false))) {
+            return false;
+        }
+
+        let dehydratedDeviceResp;
+        try {
+            dehydratedDeviceResp = await this.http.authedRequest<IDehydratedDeviceResp>(
+                Method.Get,
+                "/dehydrated_device",
+                undefined,
+                undefined,
+                {
+                    prefix: UnstablePrefix,
+                },
+            );
+        } catch (error) {
+            const err = error as MatrixError;
+            // We ignore M_NOT_FOUND (there is no dehydrated device, so nothing
+            // us to do) and M_UNRECOGNIZED (the server does not understand the
+            // endpoint).  We pass through any other errors.
+            if (err.errcode === "M_NOT_FOUND" || err.errcode === "M_UNRECOGNIZED") {
+                this.logger.info("dehydration: No dehydrated device");
+                return false;
+            }
+            throw err;
+        }
+
+        this.logger.info("dehydration: device found");
+
+        const rehydratedDevice = await this.olmMachine
+            .dehydratedDevices()
+            .rehydrate(
+                this.key!,
+                new RustSdkCryptoJs.DeviceId(dehydratedDeviceResp.device_id),
+                JSON.stringify(dehydratedDeviceResp.device_data),
+            );
+
+        this.logger.info("dehydration: device rehydrated");
+
+        let nextBatch: string | undefined = undefined;
+        let toDeviceCount = 0;
+        let roomKeyCount = 0;
+        for (;;) {
```

Comment:
Nit: I'm not sure if this is a written convention, but we normally spell this:

```suggestion
        while (true) {
```

Reply:
I think I tried using `while (true)` in some code a long time ago, and one of the linters complained about it.  I'll see if that's still the case.

Reply:
Indeed, eslint says "error  Unexpected constant condition  no-constant-condition"

Reply:
Looks like `while (true)` is used elsewhere in the code base, and they just disable the eslint warning.

Reply:
somewhat amusing that the lint rule doesn't catch `for` loops.